src/util/simplify_expr_floatbv.cpp

```diff
@@ -142,7 +141,9 @@ bool simplify_exprt::simplify_floatbv_typecast(exprt &expr)
 {
   // These casts usually reduce precision, and thus, usually round.
 
-  assert(expr.operands().size()==2);
+  PRECONDITION(expr.id() == ID_floatbv_typecast);
+  DATA_INVARIANT(
+    expr.operands().size() == 2, "float typecast has two operands");
```

Comment:
This function should take a `floatbv_typecast_exprt` as a parameter and should not be doing those checks in here.

Reply:
Note that currently all of the `simplify_XXX` work like this (take an exprt& and do a check inside). We could change that but if we were to do that I'd change all of them at once and that's out of scope for this PR IMHO.

Reply:
> [...] currently all of the [...]

Not quite true - there is sort of a gradual migration happening. Take a look at the declarations in `simplify_expr_class.h`.

Reply:
Hm... I don't really like that. It ends up violating structural invariants to still use the input expression as an out parameter, because most simplifications will "return" a different type than they were given... I find the idea of returning something that's _not_ a `plus_exprt` through a `plus_exprt` reference a bit disturbing.

Reply:
See https://github.com/diffblue/cbmc/pull/2910#discussion_r215971488.

Reply:
@tautschnig Can you take a look at https://github.com/diffblue/cbmc/pull/2939 and tell me if that's roughly in line with what you were thinking? Would you be OK with going ahead with this PR and doing further refactorings in one or more separate PRs?

Reply:
@tautschnig  I'm going to split this part out - and make a note of it to be considered in the context of #2939  and your remarks on https://github.com/diffblue/cbmc/pull/2910#discussion_r215971488